src/trusted/service_runtime/nacl_syscall_common.c

```diff
@@ -4330,3 +4320,104 @@ int32_t NaClSysSelect (struct NaClAppThread *natp, int nfds, fd_set * readfds,
   
   return retval;
 }
+
+int32_t NaClSysSigaction(struct NaClAppThread *natp, int32_t sig, const struct nacl_abi_sigaction *act, struct nacl_abi_sigaction *oact) {
+  int32_t ret;
+  struct NaClApp *nap = natp->nap;
+  const struct nacl_abi_sigaction *sysact = NULL;
+  struct nacl_abi_sigaction *sysoact = NULL;
+
+  if (act) {
+    sysact = (const struct nacl_abi_sigaction*) NaClUserToSysAddrRangeProt(nap, (uintptr_t) act, sizeof(struct nacl_abi_sigaction), NACL_ABI_PROT_READ);
+
+    if ((void*) kNaClBadAddress == sysact) {
+        NaClLog(2, "NaClSysSigaction could not translate act, returning %d\n", -NACL_ABI_EFAULT);
+        return -NACL_ABI_EFAULT;
+    }
+  }
+
+  if (oact) {
+    sysoact = (struct nacl_abi_sigaction*) NaClUserToSysAddrRangeProt(nap, (uintptr_t) oact, sizeof(struct nacl_abi_sigaction), NACL_ABI_PROT_WRITE);
+
+    if ((void*) kNaClBadAddress == sysoact) {
+        NaClLog(2, "NaClSysSigaction could not translate oact, returning %d\n", -NACL_ABI_EFAULT);
+        return -NACL_ABI_EFAULT;
+    }
+  }
+
+  ret = lind_sigaction(sig, sysact, sysoact, nap->cage_id);
+  return ret;
+}
+
+int32_t NaClSysKill(struct NaClAppThread *natp, int32_t targetcageid, int32_t sig) {
+  struct NaClApp *nap = natp->nap;
+  return lind_kill(targetcageid, sig, nap->cage_id);
+}
+
+int32_t NaClSysSigprocmask(struct NaClAppThread *natp, int32_t how, const uint64_t *nacl_set, uint64_t *nacl_oldset) {
+  int32_t ret;
+  struct NaClApp *nap = natp->nap;
+  const uint64_t *sysset = NULL;
+  uint64_t *sysoldset = NULL;
+
+  if (nacl_set) {
+    sysset = (const uint64_t *) NaClUserToSysAddrRangeProt(nap, (uintptr_t) nacl_set, sizeof(uint64_t), NACL_ABI_PROT_READ);
+
+    if ((void *) kNaClBadAddress == sysset) {
+        NaClLog(2, "NaClSysSigprocmask could not translate set, returning %d\n", -NACL_ABI_EFAULT);
+        return -NACL_ABI_EFAULT;
+    }
+  }
+
+  if (nacl_oldset) {
+    sysoldset = (uint64_t *) NaClUserToSysAddrRangeProt(nap, (uintptr_t) nacl_oldset, sizeof(uint64_t), NACL_ABI_PROT_READ);
+
+    if ((void *) kNaClBadAddress == sysoldset) {
+        NaClLog(2, "NaClSysSigprocmask could not translate oldset, returning %d\n", -NACL_ABI_EFAULT);
+        return -NACL_ABI_EFAULT;
+    }
+  }
+
+  ret = lind_sigprocmask(how, sysset, sysoldset, nap->cage_id);
+  return ret;
+}
```

Comment:
Same Functionality Refactoring: Break down the function into smaller utility functions if certain patterns are repeating. For example:
`int32_t NaClSysSigprocmask(struct NaClAppThread *natp, int32_t how, const uint64_t *nacl_set, uint64_t *nacl_oldset) {
    struct NaClApp *nap = natp->nap;
    const uint64_t *sysset = NaClTranslateAddress(nap, nacl_set, "set");
    uint64_t *sysoldset = (uint64_t *) NaClTranslateAddress(nap, nacl_oldset, "oldset");

    if (!sysset || !sysoldset) {
        return NACL_SIGPROCMASK_FAILURE;
    }

    return lind_sigprocmask(how, sysset, sysoldset, nap->cage_id);
}

const void *NaClTranslateAddress(struct NaClApp *nap, const void *addr, const char *name) {
    const void *sys_addr = NaClUserToSysAddrRangeProt(nap, (uintptr_t) addr, sizeof(uint64_t), NACL_ABI_PROT_READ);
    
    if (NACL_SIGPROCMASK_BAD_ADDRESS == sys_addr) {
        NaClLog(2, "NaClSysSigprocmask could not translate %s, returning %d\n", name, NACL_SIGPROCMASK_FAILURE);
        return NULL;
    }

    return sys_addr;
}`